### PR TITLE
sm/templates

### PR DESCRIPTION
**Pull request by sahithimuthyala**
Templates- editProject page.

### CODE REVIEW

**Comment by arpanlaha** [comment]
Cat Gif

File: api/routes/template.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+// edit draft of template by ID (what draftjs will do for template draft)
+router.put("/:template_ID/draft", function(req, res) {
+  const db = req.db;
+  const collection = db.get("templates");
+  const { template_ID } = req.params;
+  const draft = req.body.draft;
+
+  if (draft === undefined) {
+    res.sendStatus(400);
```
Comment (arpanlaha) [inline api/routes/template.js:103]: you'll want to return in this case also so that the `findOneAndUpdate` call doesnt occur

File: api/routes/template.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+// edit name of template by ID (what draftjs will do for name draft)
+router.put("/:template_ID/name", function(req, res) {
+  const db = req.db;
+  const collection = db.get("templates");
+  const { template_ID } = req.params;
+  const name = req.body.name;
+
+  if (name === undefined) {
+    res.sendStatus(400);
```
Comment (arpanlaha) [inline api/routes/template.js:129]: same as above

File: api/routes/template.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+// edit phases of template by ID (what draftjs will do for phase selectors)
+router.put("/:template_ID/phases", function(req, res) {
+  const db = req.db;
+  const collection = db.get("templates");
+  const { template_ID } = req.params;
+  const phases = req.body.phases;
+
+  if (phases === undefined) {
+    res.sendStatus(400);
```
Comment (arpanlaha) [inline api/routes/template.js:155]: same as above

File: client/components/TemplateDraft.jsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+  const [prevContent, setPrevContent] = useState(null);
+  const [editorContent, setEditorContent] = useState(null);
+
+  const saveInterval = 1000;
+  const debounceSave = (id_, json) => {
+    saveTemplateDraft(id_, json);
+    setUnsaved(false);
+  };
+  const saveCallback = useCallback(debounce(debounceSave, saveInterval), []);
```
Comment (arpanlaha) [inline client/components/TemplateDraft.jsx:35]: was this taken from brandon's stuff? i'm a bit confused as to why it needs to be a useCallback

File: client/components/TemplateDraft.jsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+      const blob = await fetch(url).then(r => r.blob());
+      const imageRef = storageRef.child(`${id}/${block.key}`);
+
+      await imageRef.put(blob).then(async function(snapshot) {
+        await snapshot.ref.getDownloadURL().then(function(url) {
+          block.data.url = url;
+        });
+      });
+    }
```
Comment (arpanlaha) [inline client/components/TemplateDraft.jsx:71]: since you're `await`ing these, you don't need to use `.then`, and can put things sequentially as if they were all synchronous methods

File: client/pages/editTemplate.jsx [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+export default function EditTemplate() {
+  const [name, setName] = useState("");
+  const [mounted, setMounted] = useState(false);
+  const [templateID, setTemplateID] = useState("");
+  const [error, setError] = useState(false);
+  const [templates, setTemplates] = useState([]);
+  const [isInspiration, setInspiration] = useState(false);
+  const [isIdeation, setIdeation] = useState(false);
+  const [isImplementation, setImplementation] = useState(false);
```
Comment (arpanlaha) [inline client/pages/editTemplate.jsx:31]: would it make more sense to combine these into a single state? maybe an object with keys for the phases or an array containing the selected phases?

File: client/utils/apiWrapper.js [hunk 7]
```diff
@@ -736,6 +737,133 @@ export const getComments = model_id => {
...
+
+export const deleteTemplate = Template_ID => {
+  /**
+   * Deletes a template
+   * Returns DELETE_TEMPLATE_FAIL upon failure
+   */
+  const requestString = `${BASE_URL}/templates/${Template_ID}`;
+  return axios
+    .delete(requestString, {
+      headers: {}
```
Comment (arpanlaha) [inline client/utils/apiWrapper.js:769]: can remove headers since it's empty

File: client/components/TemplateDraft.jsx (hunk 5 above)
Comment (sahithimuthyala) [reply client/components/TemplateDraft.jsx:71]: Can I do the same on line 63?
`const blob = await fetch(url).then((r) => r.blob());`

Comment (arpanlaha) [reply client/components/TemplateDraft.jsx:71]: yup you should be able to!

File: client/components/TemplateDraft.jsx (hunk 4 above)
Comment (sahithimuthyala) [reply client/components/TemplateDraft.jsx:35]: Yes, Ashwin took this part from Brandon's stuff in components/Draft.jsx. Looks like it doesn't need to be a useCallback though. Line 35 modified to:
`const saveCallback = debounce(debounceSave, saveInterval);`

File: client/pages/editTemplate.jsx (hunk 6 above)
Comment (sahithimuthyala) [reply client/pages/editTemplate.jsx:31]: Left as separate states.

File: api/public/schema/userSchema.js [hunk 8]
```diff
@@ -1,5 +1,5 @@
 const schema = {
-  id: "/UserSchema",
+  id: "/TemplateSchema",
```
Comment (yousefa00) [inline api/public/schema/userSchema.js:2]: Why is UserSchema being renamed here?

File: api/routes/template.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+      }
+    });
+  }
+);
+
+// delete template by ID (delete button)
+router.delete("/:template_ID", function(req, res) {
+  // CHECK AUTH
+  const db = req.db;
+  let id = req.params.template_ID;
```
Comment (yousefa00) [inline api/routes/template.js:36]: Should be const

File: api/routes/template.js [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+  const collection = db.get("templates");
+  collection.find({}, {}, function(e, docs) {
+    res.send(docs);
+  });
+});
+
+// get specific template by id (clicking on template in dropdown menu)
+router.get("/:template_ID", function(req, res) {
+  const db = req.db;
+  let id = req.params.template_ID;
```
Comment (yousefa00) [inline api/routes/template.js:62]: here as well

File: api/routes/user.js [hunk 11]
```diff
@@ -8,6 +8,23 @@ const SUCCESS = 200;
 const NOT_FOUND = 404;
 const DUPLICATE = 409;
 
+router.get("/", checkToken, (req, res) => {
```
Comment (yousefa00) [inline api/routes/user.js:11]: We cant have two get endpoints with the same route. Could you remove this one, please?

File: client/components/Head.jsx [hunk 12]
```diff
@@ -14,8 +14,8 @@ export default function Head(props) {
       <link
         rel="stylesheet"
         href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
-        integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T"
-        crossOrigin="anonymous"
+        // integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T"
```
Comment (yousefa00) [inline client/components/Head.jsx:17]: Why are these commented out?

File: client/components/TemplateDraft.jsx [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        setLoading(false);
+      })
+      .catch(() => {
+        setLoading(false);
+      });
+  }, [id]);
+
+  const renderDraft = () => {
+    if (loading) {
+      return undefined;
```
Comment (yousefa00) [inline client/components/TemplateDraft.jsx:104]: Use null instead.

File: client/pages/projects/index.jsx [hunk 14]
```diff
@@ -9,6 +9,7 @@ import levelData from "../../utils/levels";
 import { getModels } from "../../utils/apiWrapper";
 import {
   Card,
+  CardImg,
```
Comment (yousefa00) [inline client/pages/projects/index.jsx:12]: Please undo the changes in this file.

File: client/public/styles/editTemplate.scss [hunk 15]
```diff
@@ -0,0 +1,126 @@
+body {
+  background-color: #eee;
+  font-family: Arvo, serif;
+  font-weight: bold;
+  color: #003366;
+}
+.edit-template-div {
+  background-color: white;
+  text-align: center;
+  width: 836px;
```
Comment (yousefa00) [inline client/public/styles/editTemplate.scss:10]: We want to use % for width values, generally.

File: client/public/styles/editTemplate.scss [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  height: 100%;
+  margin-left: auto;
+  margin-right: auto;
+  //   margin-bottom: 20%;
+}
+.bottom-buttons {
+  width: 100%;
+  margin-top: 10px;
```
Comment (yousefa00) [inline client/public/styles/editTemplate.scss:18]: Here, you can use rem or %

Comment (yousefa00) [reply client/public/styles/editTemplate.scss:18]: And same for all the other px values in this css file.

File: client/utils/apiWrapper.js [hunk 17]
```diff
@@ -18,6 +18,7 @@ export const getModels = (sdg_query, searchPage = null) => {
   } else {
     requestString = `${BASE_URL}/models`;
   }
+  console.log(requestString);
```
Comment (yousefa00) [inline client/utils/apiWrapper.js:21]: log

File: client/utils/apiWrapper.js [hunk 18]
```diff
@@ -736,6 +737,129 @@ export const getComments = model_id => {
   }));
 };
 
+// TEMPLATE WRAPPERS -----------
```
Comment (yousefa00) [inline client/utils/apiWrapper.js:740]: You can remove this comment

File: client/utils/apiWrapper.js [hunk 19]
```diff
@@ -736,6 +737,129 @@ export const getComments = model_id => {
...
+    })
+    .catch(error => {
+      ({
+        type: "POST_TEMPLATE_FAIL",
+        error
+      });
+    });
+};
+
+export const deleteTemplate = Template_ID => {
```
Comment (yousefa00) [inline client/utils/apiWrapper.js:761]: we should use camel case for variable naming

File: client/utils/apiWrapper.js [hunk 20]
```diff
@@ -736,6 +737,129 @@ export const getComments = model_id => {
...
+    })
+    .catch(error => {
+      ({
+        type: "GET_TEMPLATE_FAIL",
+        error
+      });
+    });
+};
+
+export const getTemplateByID = Template_ID => {
```
Comment (yousefa00) [inline client/utils/apiWrapper.js:795]: same as above

File: client/pages/editTemplate.jsx [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+
+  const phases = ["Inspiration", "Ideation", "Implementation"];
+  const implementationIndex = 2;
```
Comment (yousefa00) [inline client/pages/editTemplate.jsx:34]: I know that inspiration and ideation indices aren't flagged by the linter because they're 0 and 1, but if we're going to do this, we should add globals for those two as well. I'll probably change it later to be one state with an object.

File: client/pages/editTemplate.jsx [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+        if (!mounted) {
+          setTemplateID(currentTemplates.data[0]._id);
+          setMounted(true);
+
+          result = await getTemplateByID(currentTemplates.data[0]._id);
+        } else {
+          result = await getTemplateByID(templateID);
+        }
+
+        if (result.data !== undefined && result) {
```
Comment (yousefa00) [inline client/pages/editTemplate.jsx:58]: These should be flipped so that they are being checked in the right order.

File: client/pages/editTemplate.jsx [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+            setInspiration(false);
+            setIdeation(false);
+            setImplementation(false);
+          }
+        }
+      }
+    };
+
+    setDisplay();
+  }, [templateID]);
```
Comment (yousefa00) [inline client/pages/editTemplate.jsx:80]: Do you need to include these dependencies from the lint warning?

File: client/pages/editTemplate.jsx [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+
+  const handleID = clickedTemplateID => {
+    setTemplateID(clickedTemplateID);
+  };
```
Comment (yousefa00) [inline client/pages/editTemplate.jsx:84]: You might not need this, see comment a little bit below.

File: client/pages/editTemplate.jsx [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+
+  const handleNewStage = async e => {
+    e.preventDefault();
```
Comment (yousefa00) [inline client/pages/editTemplate.jsx:87]: Why do you need to prevent the default action of the event? Might not need this.

File: client/pages/editTemplate.jsx [hunk 26]
```diff
@@ -0,0 +1,247 @@
...
+              </Button>
+            </Row>
+            <Row>
+              {templates.data &&
+                templates.data.map(template => (
+                  <Button
+                    key={template._id}
+                    className="stage-button"
+                    color="white"
+                    onClick={() => handleID(template._id)}
```
Comment (yousefa00) [inline client/pages/editTemplate.jsx:159]: here, you may be able to just have `onClick={handleID}`, try this and lmk.

File: client/pages/editTemplate.jsx (hunk 21 above)
Comment (sahithimuthyala) [reply client/pages/editTemplate.jsx:34]: Sounds good, I added globals for the inspiration and ideation indices for now.

File: client/pages/editTemplate.jsx (hunk 23 above)
Comment (sahithimuthyala) [reply client/pages/editTemplate.jsx:80]: I don't think so. I added `mount` though, but if I add `phases`, then the template name and phases are laggy when switching between templates.

File: client/public/styles/editTemplate.scss (hunk 15 above)
Comment (sahithimuthyala) [reply client/public/styles/editTemplate.scss:10]: Issue when the width is a % for .edit-template-div:
Screen Shot 2020-05-28 at 8 21 31 PM
